packages/host/app/services/operator-mode-state-service.ts

```diff
@@ -462,15 +497,75 @@ export default class OperatorModeStateService extends Service {
         } else {
           dirs.splice(i, 1);
         }
-        this.state.openDirs = dirs;
+        this.openDirs.set(this.realmURL, new TrackedArray(dirs));
         return;
       } else if (entryPath.startsWith(dirs[i])) {
         dirs[i] = entryPath;
-        this.state.openDirs = dirs;
+        this.openDirs.set(this.realmURL, new TrackedArray(dirs));
         return;
       }
     }
-    this.state.openDirs = [...dirs, entryPath];
+    this.openDirs.set(this.realmURL, new TrackedArray([...dirs, entryPath]));
     this.schedulePersist();
   }
+
+  private get readyFile() {
+    if (isReady(this.openFile.current)) {
+      return this.openFile.current;
+    }
+    throw new Error(
+      `cannot access file contents ${this.state.codePath} before file is open`,
+    );
+  }
+
+  get openFileIsReady() {
+    return isReady(this.openFile.current);
+  }
+
+  get realmURL() {
+    if (isReady(this.openFile.current)) {
+      return this.readyFile.realmURL;
+    } else if (this.cachedRealmURL) {
+      return this.cachedRealmURL.href;
+    }
+
+    return this.cardService.defaultURL.href;
```

Comment:
I think we discussed last week on the call to refrain from using `defaultURL`. Either we have the realm url from the file or we don't, there's no default to fall back to - that's my understanding.

Reply:
it’s true and I did try removing it but there are tests like [this](https://github.com/cardstack/boxel/blob/97a8eacb72f858330461611f6a59b41564e6415a/packages/host/tests/acceptance/code-mode-test.ts#L1013) that rely on the default realm. I think it would be good to finally stop using this property but it’ll have to happen in future work.